research/05-problem-ana.tex

```diff
@@ -1,10 +1,127 @@
 \section{Problem Analysis}
 \label{sec:problem-analysis}
+This section gives an analysis of the problem as defined in the
+previous section \ref{sec:problem-definition}. More specifically, this
+section describes the problems that are expected and to be solved
+during the implementation process of the product. The first part of
+this section describes the process of incorporating REPL generation
+within the larger context of Spoofax. The second part analyzes the
+part of the problem concerned with literate programming, but this part
+will be analyzed in less detail than generating REPLs as it is unsure
+whether this will be present in the final product.
 
-\subsection{How does REPL generation fit within Spoofax}
-\label{ssec:architecture}
+\subsection{Incorporating REPL generation within Spoofax}
+\label{ssec:incorp-repl-gener}
+To generate solidly working REPLs for any language defined in Spoofax,
+some Spoofax concepts have to be carefully considered. Since every
+language can have many different language constructs, generating REPLs
+should not be constrained to language specific constructs like
+classes, but should rather be done generically. Therefore the
+implementation and particularly the interaction with the user needs to
+be thought of carefully.
 
-\subsection{How are plugins developed for Eclipse?}
+This section lists the problems that are expected when trying to
+implement generating REPLs generically. For each problem, their
+solution approaches that should be explored during the implementation
+process are given, as well as the parts of Spoofax that are relevant
+for that problem.
+
+\subsubsection{Detecting unfinished expressions for multiline editing}
+\label{sec:detect-unfin-expr}
+To support multiline editing, the generated REPL should detect that
+the expression is unfinished when the user presses ``enter'' for the
+next line, instead of trying to parse and execute the unfinished
+expression. An obvious part of Spoofax that is relevant for this
+problem, is the syntax definition in SDF3 (\cref{sec:orgheadline1}).
+
+\subsubsection{Language specific additional commands}
+\label{sec:lang-spec-addit}
+Another problem is when some additional command can be useful for a
+REPL of a particular language, but would not make any sense within the
+context of other languages. For example, some languages such as Java
+have the notion of a ``module'' or ``package'', which declares a
+separate namespace for the definitions in it. An additional command to
+list all of the currently loaded modules is useful, but would not make
+any sense in languages that have no concept of a module.
```

Comment:
Aren't all these things defined in their separate namespaces, whatever there may be? **All** languages in Spoofax use NaBL for their name bindings, so **all** languages in Spoofax have these namespaces that we can list. As such, I think the example issue given here does not make much sense. Even more so, because **all** languages have these namespaces, we can just add a language-agnostic command enabled in **all** REPLs to list these namespaces. That is precisely what the feature "Additional commands to the REPL" means.